### PR TITLE
Cart

### PR DESCRIPTION
**Pull request by Tramchandani**
Screenshot from 2021-03-24 17-51-50
Screenshot from 2021-03-24 17-51-38
Screenshot from 2021-03-24 17-51-34
Screenshot from 2021-03-24 17-51-26

### CODE REVIEW

**Comment by kalpana19sharma** [comment]
@Tramchandani please resolve the conflicts.

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  before_action :current_user
+  before_action :current_cart
+
+  def index
+    @exercises = @cart.exercises
+    render layout: 'empty'
+  end
+
+  def add_to_cart
+    exercise = Exercise.find_by_id(exercise_id_params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/carts_controller.rb]: Move this logic before filter

File: app/controllers/carts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  private
+
+  def current_cart
+    if @user.cart.nil?
+      @cart = @user.create_cart
+    else
+      @cart = @user.cart
+    end
```
Comment (kalpana19sharma) [inline app/controllers/carts_controller.rb]:
```suggestion
    @cart = @user.cart || @user.create_cart
```

File: app/controllers/carts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+class CartsController < ApplicationController
+  before_action :current_user
+  before_action :current_cart
```
Comment (kalpana19sharma) [inline app/controllers/carts_controller.rb]: Rename these methods

File: app/controllers/carts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    if @user.cart.nil?
+      @cart = @user.create_cart
+    else
+      @cart = @user.cart
+    end
+  end
+
+  def current_user
+    @user ||= User.find_by_id(session[:user_id])
+  end
```
Comment (kalpana19sharma) [inline app/controllers/carts_controller.rb]: I thought you had this method in application controller. And do we really need to add `before_action` filter for this?

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -11,6 +11,16 @@ def check_user?
     session[:user_id] && User.find_by_id(session[:user_id])
   end
 
+  def in_cart?(exercise)
+    user = User.find_by_id(session[:user_id])
+    if user.cart.nil?
+      cart = user.create_cart
+    else
+      cart = user.cart
+    end
```
Comment (kalpana19sharma) [inline app/helpers/application_helper.rb]: You can eliminate this logic by adding a after create callback in user model.

File: app/models/exercise_in_cart.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class ExerciseInCart < ApplicationRecord
```
Comment (kalpana19sharma) [inline app/models/exercise_in_cart.rb]: Rename this model to `CartItem`

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -11,6 +11,16 @@ def check_user?
     session[:user_id] && User.find_by_id(session[:user_id])
   end
 
+  def in_cart?(exercise)
```
Comment (kalpana19sharma) [inline app/helpers/application_helper.rb]: Can we move this method in model?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/carts_controller.rb (hunk 4 above)
Comment (Tramchandani) [reply app/controllers/carts_controller.rb]: yes ma'am it is already in application controller so it is redundant here. I will remove it.

File: app/models/exercise_in_cart.rb (hunk 6 above)
Comment (Tramchandani) [reply app/models/exercise_in_cart.rb]: should I rename database table as well?

Comment (kalpana19sharma) [reply app/models/exercise_in_cart.rb]: yes

File: app/controllers/carts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  def user_cart
+    @cart = current_user.cart
+  end
+
+  def exercise_id_params
+    params.permit(:id)
+  end
+
+  def selected_exercise
+    @exercise = Exercise.find_by_id(exercise_id_params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/carts_controller.rb:39]: What if exercise is not found?

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -8,12 +8,16 @@ def show_errors(object, field)
...
 
   def full_address(address)
     @full_address = address.house_number.to_s.concat(", ").concat(address.colony).concat(" ")
     @full_address.concat(address.street).concat(" ").concat(address.city).concat(", ")
     @full_address.concat(address.state)
   end
+
+  def current_user
+    User.find_by_id(session[:user_id])
+  end
```
Comment (kalpana19sharma) [inline app/helpers/application_helper.rb]: You already have this method in application controller so better declare that as helper.

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/carts_controller.rb (hunk 8 above)
Comment (Tramchandani) [reply app/controllers/carts_controller.rb:39]: when user selects the exercise the exercise is added to cart and user is redirected on the same page.
SO it is not possible to change the exercise id through url.

**Comment by kalpana19sharma** [comment]
@Tramchandani please resolve the conflicts.

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**